### PR TITLE
Add `.TextColor()` and `.Text()` Extension Methods

### LINKED ISSUE #38: [Proposal] Add `.Text()` and `.TextColor()` Extension Methods
# Add `.Text()` + `.TextColor()` Extension Methods

* [x] Proposed
* [x] Prototype
* [x] Implementation
  * [x] iOS Support
  * [x] Android Support
  * [x] macOS Support
  * [x] Windows Support
* [x] Unit Tests
* [x] Sample
* [x] Documentation

## Link to Discussion

https://github.com/CommunityToolkit/Maui.Markup/discussions/37

## Summary
[summary]: #summary

This Proposal adds extension methods for `IText.Text` and `ITextStyle.TextColor`.

## Motivation
[motivation]: #motivation

Currently, the only way to set `Text` and `TextColor` properties on an element that implements `IText` is to set the properties directly. Here is an example from `CommunityToolkit.Maui.Markup.Sample.Pages.SettingsPage`:

```cs
new Label { Text = "Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor }
	.LayoutFlags(AbsoluteLayoutFlags.XProportional | AbsoluteLayoutFlags.WidthProportional)
	.LayoutBounds(0, 0, 1, 40)
	.TextCenterHorizontal()
	.TextBottom(),
```

This Proposal will allow devs to use fluent extension methods to set both `Text` and `TextColor`.

## Detailed Design
[design]: #detailed-design

```cs
	/// <summary>
	/// Sets <see cref="ITextStyle.TextColor"/> Property
	/// </summary>
	/// <typeparam name="TBindable"><see cref="BindableObject"/></typeparam>
	/// <param name="bindable">Element</param>
	/// <param name="textColor">Text <see cref="Color"/></param>
	/// <returns></returns>
	public static TBindable TextColor<TBindable>(this TBindable bindable, Color? textColor) where TBindable : BindableObject, ITextStyle
	{
		bindable.SetValue(TextElement.TextColorProperty, textColor);
		return bindable;
	}

	/// <summary>
	/// Sets <see cref="IText.Text"/> Property
	/// </summary>
	/// <typeparam name="TBindable"><see cref="BindableObject"/></typeparam>
	/// <param name="bindable">Element</param>
	/// <param name="text"></param>
	/// <returns></returns>
	public static TBindable Text<TBindable>(this TBindable bindable, string? text) where TBindable : BindableObject, IText
	{
		bindable.SetValue(ITextElement.TextProperty, text);
		return bindable;
	}

	/// <summary>
	/// Sets <see cref="IText.Text"/> Property
	/// </summary>
	/// <typeparam name="TBindable"><see cref="BindableObject"/></typeparam>
	/// <param name="bindable">Element</param>
	/// <param name="text"></param>
	/// <param name="textColor">Text <see cref="Color"/></param>
	public static TBindable Text<TBindable>(this TBindable bindable, string? text, Color? textColor) where TBindable : BindableObject, IText
	{
		return bindable.Text(text).TextColor(textColor);
	}
```

## Usage Syntax
[usage]: #usage-syntax

### C# Usage

```cs
Content = new VerticalStackLayout
{
  Children = 
  {
    new Label()
      .Text("Hello World", Colors.Blue),

    new Label()
      .TextColor(Colors.Green)
      .Bind(Label.TextProperty, nameof(ViewModel.LabelText);
  }
}

```

## Drawbacks
[drawbacks]: #drawbacks

Currently, [`TextElement` does not have a `TextProperty` static property](https://github.com/brminnick/maui/blob/main/src/Controls/src/Core/TextElement.cs), i.e. `TextElelement.TextProperty` does not yet exist.

We can workaround this by doing the following. This is a proven and common workaround in [Xamarin.CommunityToolkit.Markup](https://github.com/xamarin/XamarinCommunityToolkit/blob/f3663cd6d3f980b86b8cfc067123d81708db4abe/src/Markup/Xamarin.CommunityToolkit.Markup/ElementExtensions.cs#L3-L4).

```cs
using ITextElement = Microsoft.Maui.Controls.Label; // ToDo Remove this once TextElement.TextProperty is added

public static class ElementExtensions
{
	public static TBindable Text<TBindable>(this TBindable bindable, string? text) where TBindable : BindableObject, IText
	{
		bindable.SetValue(ITextElement.TextProperty, text);
		return bindable;
	}
}
```

## Alternatives
[alternatives]: #alternatives

Currently, the only alternative is to initialize the properties like so:
```cs
new Label { Text = "Top Stories To Fetch", TextColor = ColorConstants.PrimaryTextColor }
```

## Unresolved Questions
[unresolved]: #unresolved-questions

Should we support `null` inputs? The default value of `Color` and `Text` is `null` and I imagine the number of use cases where developer would use this `.Text()` to revert the values of `Color` and `Text` back to `null` may be minimal.

### PR DESCRIPTION
**Pull request by brminnick**
### Description of Change ###

This Pull Request implements the `.TextColor()` and `.Text()` extension methods to `CommunityToolkit.Maui.Markup`.

 ### Linked Issues ###
 

 - Fixes #38 

 ### PR Checklist ###
 - [x] Has a linked Issue, and the Issue has been `approved`(bug) or `Championed` (feature/proposal)
 - [x] Has tests (if omitted, state reason in description)
 - [x] Has samples (if omitted, state reason in description)
 - [x] Rebased on top of `main` at time of PR
 - [x] Changes adhere to [coding standard](https://github.com/CommunityToolkit/Maui.Markup/blob/main/CONTRIBUTING.md#contributing-code---best-practices)
 - [x] Documentation created or updated: https://github.com/MicrosoftDocs/CommunityToolkit/pull/47

 ### Additional information ###

This PR also updates Unit Tests and the Sample App to include `.TextColor()` and `.Text()`.

### CODE REVIEW

File: src/CommunityToolkit.Maui.Markup.UnitTests/ElementExtensionsTests.cs [hunk 1]
```diff
@@ -50,39 +50,63 @@ public void EffectsMultiple()
...
+	[Test]
+	public void Text_ProvidedColor()
+		=> TestPropertiesSet(l => l.Text("Hello World", Colors.Green), (Label.TextProperty, "Hello World"), (TextElement.TextColorProperty, Colors.Green));
+
+	[Test]
+	public void Text_CustomColor()
+		=> TestPropertiesSet(l => l.Text("Hello World", new Color(250, 5, 128, 1)), (Label.TextProperty, "Hello World"), (TextElement.TextColorProperty, new Color(250, 5, 128, 1)));
+
+	[Test]
+	public void Text_NulValues()
```
Comment (bijington) [inline src/CommunityToolkit.Maui.Markup.UnitTests/ElementExtensionsTests.cs:108]: Minor typo

```suggestion
	public void Text_NullValues()
```

**Review by bijington — APPROVED** [review]
Impressive turnaround time!!

Just the typo I noticed

**Comment by brminnick** [comment]
Looks like I found a bug in .NET MAUI.

`ToolbarItem` implements `IText`, but it doesn't have a `TextProperty` bindable object.

This means that the following code crashes:
```cs
new ToolbarItem().Text("Settings");
```

I'll raise this to the .NET MAUI team.

**Comment by brminnick** [comment]
Docs PR Opened: https://github.com/MicrosoftDocs/CommunityToolkit/pull/47

**Pull request merged**